src/storage/sqlite/mod.rs

```
@@ -0,0 +1,428 @@
+                Ok(storage) => {
+                    // Send Ok back to the caller of `new` and then start the actor loop.
+                    init_sender.send(Ok(())).unwrap();
+                    let mut actor = Actor { storage, receiver };
+                    actor.run();
+                }
+                Err(e) => {
+                    // Send the initialization error back.
+                    init_sender.send(Err(e)).unwrap();
+                }
+            }
+        });
+
+        // Block until the thread sends its initialization result.
+        init_receiver.recv().unwrap()?;
+        Ok(Self { sender })
+    }
+}
+
+#[async_trait]
+impl Storage for SqliteStorage {
+    // Sends the BeginTxn message to the underlying SqliteStorage. Now that the txn has
+    // begun, this async txn obj can be passed around and operated upon, as it
+    // communicates with the underlying sync txn.
+    async fn txn<'a>(&'a mut self) -> Result<Box<dyn StorageTxn + Send + 'a>> {
+        let (reply_tx, reply_rx) = oneshot::channel();
+        self.sender
+            .send(ActorMessage::BeginTxn(reply_tx))
+            .map_err(|e| Error::Other(e.into()))?;
+        let txn_sender = reply_rx.await??;
+        Ok(Box::new(ActorTxn::new(txn_sender)))
+    }
+}
+
+/// An async proxy for a transaction running on the sync actor thread.
+pub(super) struct ActorTxn {
+    sender: mpsc::Sender<TxnMessage>,
+    committed: bool,
+}
+
+impl ActorTxn {
+    fn new(sender: mpsc::Sender<TxnMessage>) -> Self {
+        Self {
+            sender,
+            committed: false,
+        }
+    }
+
+    async fn call<R, F>(&self, f: F) -> Result<R>
+    where
+        F: FnOnce(oneshot::Sender<Result<R>>) -> TxnMessage,
+        R: Send + 'static,
+    {
+        let (tx, rx) = oneshot::channel();
+        self.sender
+            .send(f(tx))
+            .map_err(|e| Error::Other(e.into()))?;
+        rx.await?
```

This async call could be cancelled here, and I wanted to think through what would happen in that case.

The `rx` would be dropped meaning an error in the actor thread when the query is complete. But that's ignored with `let _ = ..`. So it goes around the loop again waiting for another message. If the ActorTxn is not dropped, this is fine -- it will either send another query or commit/rollback.

If the ActorTxn is also dropped when this call is cancelled, then its `drop` method will get called, so when the current query completes, the actor will see and ignore an error, then go around the loop again and receive the Rollback message.

All of that is well and good, but I figured it'd be worth recording the thought process!